File: 03.03.堆盘子/03.03.cpp
class StackOfPlates {
public:
    StackOfPlates(int cap) {
        capacity = cap;
    }
    
    void push(int val) {
        if (capacity == 0) {
            // 如果capacity为0，直接return
            return;
        }
        if (stks.empty() || stks.back().size() == capacity) {
            // 当前vector为空或当前栈满，新增一个栈
            stks.emplace_back(stack<int>());
        }
        stks.back().push(val);
    }
    
    int pop() {
        if (capacity == 0 || stks.empty()) {
            // capacity为0或当前vector为空，直接return -1
            return -1;
        }
        int res = stks.back().top();
        stks.back().pop();
        if (stks.back().empty()) {
            // 如果最后一个栈为空，需要弹出最后一个栈
            stks.pop_back();
        }
        return res;
    }
    
    int popAt(int index) {
        if (capacity == 0 || index >= stks.size() || stks[index].empty()) {
            // capacity为0或下标越界或当前栈空，直接return -1
            return -1;
        }
        int res = stks[index].top();
        stks[index].pop();
        if (stks[index].empty()) {
            // 如果当前栈空，最后删除当前栈
            stks.erase(stks.begin() + index);
        }
        return res;
    }
private:
    vector<stack<int>> stks;
    int capacity;
};

/**
 * Your StackOfPlates object will be instantiated and called as such:
 * StackOfPlates* obj = new StackOfPlates(cap);
 * obj->push(val);
 * int param_2 = obj->pop();
 * int param_3 = obj->popAt(index);
 */

File: 03.03.堆盘子/03.03.go
type StackOfPlates struct {
	cap   int
	stack [][]int
}

func Constructor(cap int) StackOfPlates {
	return StackOfPlates{
		cap:   cap,
		stack: make([][]int, 0),
	}
}

func (this *StackOfPlates) Push(val int) {
	if this.cap == 0 {
		return
	}

	if len(this.stack) == 0 {
		newPlate := []int{val}
		this.stack = append(this.stack, newPlate)
		return
	}

	if len(this.stack[len(this.stack)-1]) == this.cap {
		newPlate := []int{val}
		this.stack = append(this.stack, newPlate)
		return
	}

	this.stack[len(this.stack)-1] = append(this.stack[len(this.stack)-1], val)
}

func (this *StackOfPlates) Pop() int {
	if len(this.stack) == 0 {
		return -1
	}

	plate := this.stack[len(this.stack)-1]
	v := plate[len(plate)-1]

	plate = plate[0 : len(plate)-1]
	this.stack[len(this.stack)-1] = plate

	if len(plate) == 0 {
		this.stack = this.stack[0 : len(this.stack)-1]
	}

	return v
}

func (this *StackOfPlates) PopAt(index int) int {
	n := len(this.stack)
	if index >= n {
		return -1
	}

	plate := this.stack[index]
	v := plate[len(plate)-1]
	plate = plate[0 : len(plate)-1]
	this.stack[index] = plate

	if len(plate) == 0 {
		tmp := this.stack[index+1:]
		this.stack = this.stack[:index]
		this.stack = append(this.stack, tmp...)
	}

	return v
}

/**
 * Your StackOfPlates object will be instantiated and called as such:
 * obj := Constructor(cap);
 * obj.Push(val);
 * param_2 := obj.Pop();
 * param_3 := obj.PopAt(index);
 */